src/main/java/org/jenkinsci/maven/plugins/hpi/AbstractHpiMojo.java

```diff
@@ -827,31 +827,36 @@ private static void copyFilteredFile(File from, File to, String encoding, Filter
     /**
      * If the project is on Git, figure out Git SHA1.
      *
-     * @return null if no git repository is found
-     */
-    public String getGitHeadSha1() {
-        // we want to allow the plugin that's not sitting at the root (such as findbugs plugin),
-        // but we don't want to go up too far and pick up unrelated repository.
-        File git = new File(project.getBasedir(), ".git");
-        if (!git.exists()) {
-            git = new File(project.getBasedir(),"../.git");
-            if (!git.exists())
-                return null;
+     * @param shortHash if {@code true} returns a short (8 character) hash, otherwise the fill object hash. 
+     * @return {@code null} if no git repository is found
+     */
+    public String getGitHeadSha1(boolean shortHash) {
+        String v = getGitRevParseOutput("HEAD");
+        if (v == null) {     // not a git repo
+            return null;
+        }
+        if (v.length()<8) {
+            return null;    // git repository present, but without commits
+        } 
+        if (shortHash) {
+            return v.substring(0,8);
+        }
+        return v;
+    }
+    
+    protected String getGitRevParseOutput(String revparseArg) {
+        if (!project.getScm().getConnection().startsWith("scm:git")) {
```

Comment:
rather than looking for a `.git` folder in the hierachy we now just ask the project if its connection starts with `scm:git`

This does not need to be set for every plugin in a reactor, as the model we have contains inherited data.